source/client/output_formatter_impl.cc

```diff
@@ -138,5 +144,123 @@ DottedStringOutputFormatterImpl::formatProto(const nighthawk::client::Output& ou
+}
+
+std::string FortioOutputFormatterImpl::formatProto(const nighthawk::client::Output& output) const {
+  nighthawk::client::FortioResult fortio_output;
+
+  // TODO(nareddyt): Not needed but nice to have, displays in the UI
+  fortio_output.mutable_labels()->set_value("");
+  fortio_output.mutable_starttime()->set_seconds(output.timestamp().seconds());
+  fortio_output.mutable_requestedqps()->set_value(output.options().requests_per_second().value());
+  fortio_output.mutable_url()->set_value(output.options().uri().value());
+
+  // Actual and requested durations are the same
+  const auto& nh_duration = output.options().duration().seconds();
+  fortio_output.mutable_requestedduration()->set_seconds(nh_duration);
+  fortio_output.mutable_actualduration()->set_value(nh_duration);
+
+  // This displays as "connections" in the UI, not threads
+  fortio_output.mutable_numthreads()->set_value(output.options().connections().value());
+
+  // Get the result that represents all workers (global)
+  const auto& nh_global_result = this->getGlobalResult(output);
+
+  // Fill in the actual QPS based on the counters
+  const auto& nh_rq_counter = this->getCounterByName(nh_global_result, "upstream_rq_total");
+  const double actual_qps = static_cast<double>(nh_rq_counter.value()) / nh_duration;
+  fortio_output.mutable_actualqps()->set_value(actual_qps);
+
+  // Fill in the number of successful responses.
+  // Fortio-ui only reads the 200 OK field, other fields are never displayed.
+  fortio_output.mutable_retcodes()->insert({"200", 0});
+  try {
+    const auto& nh_2xx_counter = this->getCounterByName(nh_global_result, "benchmark.http_2xx");
+    fortio_output.mutable_retcodes()->at("200") = nh_2xx_counter.value();
+  } catch (const NighthawkException& e) {
+    // If this field doesn't exist, then there were no 2xx responses
+    fortio_output.mutable_retcodes()->at("200") = 0;
+  }
+
+  // The core of the transformation is here: All the percentiles to display the dashboard
+  const auto& nh_stat = this->getRequestResponseStatistic(nh_global_result);
+
+  // Set the count (number of data points)
+  nighthawk::client::DurationHistogram fortio_histogram;
+  fortio_histogram.mutable_count()->set_value(nh_stat.count());
+
+  uint64_t prev_fortio_count = 0;
+  double prev_fortio_end = 0;
+  for (int i = 0; i < nh_stat.percentiles().size(); i++) {
+
+    nighthawk::client::DataEntry fortio_data_entry;
+    const auto& nh_percentile = nh_stat.percentiles().at(i);
+
+    // fortio_percent = 100 * nh_percentile
+    fortio_data_entry.mutable_percent()->set_value(nh_percentile.percentile() * 100);
+
+    // fortio_count = nh_count - prev_fortio_count
+    fortio_data_entry.mutable_count()->set_value(nh_percentile.count() - prev_fortio_count);
+
+    // fortio_end = nh_duration (need to convert formats)
+    const double nh_duration_in_double = nh_percentile.duration().nanos() / 1e9;
```

Comment:
I'm not sure this will work with larger durations, you might also need to consider the `.seconds()` part.

Reply:
Yeah, see Envoy::Protobuf::util::TimeUtil::DurationToNanoseconds. which may be useful here.

Reply:
I think I see, could you confirm I did it correctly now?